### PR TITLE
Clean merge of DB online/offline changes against latest master

### LINKED ISSUE #908: Allow an Admin to take a DB Offline
Break out this action item from #748 for allocation at the next spring planning

### LINKED ISSUE #909: Allow Admin to take an Offline DB Online
Break out this action item from #748 for allocation in the next sprint planning

### LINKED ISSUE #910: Update _resync to use bucket online/offline
Break out this action item from #748 for allocation at the next sprint planning

### LINKED ISSUE #911: Detection of lost TAP feed impacting service availability
Break out this action item from #748 for allocation at the next sprint planning

### LINKED ISSUE #912: Write unit tests for bucket online/offline updates
Break out this action item from #748 for allocation at the next sprint planning

### LINKED ISSUE #1012: Put bucket in offline mode for n seconds
After completion of #1004

Add a new configuration feature on startup, put bucket in offline mode for n seconds'

This might be useful for bringing up a cluster of SG instances, and making sure all instances of a DB come up at the same time.

### LINKED ISSUE #1013: Automatically put bucket in offline mode when lose TAP feed
After completion of #1004

Add the following feature

when we get an error on the TAP feed, automatically put bucket in offline mode

This is the wrap-up ticket for all the following related tickets
- [x] - 622   - not clear behavior with SG when sync CB bucket recreated
- [x] - 748   - Enable taking a bucket offline/online
- [x] - 907   - Document implementation design taking bucket offline/online
- [x] - 908   - Allow an Admin to take a DB Offline
- [x] - 909   - Allow Admin to take an Offline DB Online
- [x] - 910   - Update _resync to use bucket online/offline
- [x] - 911   - Detection of lost TAP feed impacting service availability
- [x] - 912   - Write unit tests for bucket online/offline updates
- [x] - 942   - Handle changes to Couchbase Server cluster topology at runtime
- [x] - 1004 - Implementation of bucket online/offline
- [x] - 1011 - Uptake BucketUpdater for upgrade, stability
- [x] - 1170 - Bucket online/offline knowledge transfer
- [x] - 1176 - Add WebHook notifications for bucket online/offline
- [x] - 1250 - Potential test scenarios for DB online/offline

### PR DESCRIPTION
**Pull request by ajres**
A clean merge of the changes for DB online/offline with latest master
Scope covers

fixes #908
fixes #909
fixes #910
fixes #911
fixes #912
fixes #1012
fixes #1013

And bring DB online after specified delay in seconds

### CODE REVIEW

**Comment by ajres** [comment]
Jenkins dev [build](http://mobile.jenkins.couchbase.com/view/Sync_Gateway_Dev/job/sync-gateway-dev-build/368/)

File: src/github.com/couchbase/sync_gateway/base/bucket.go [hunk 1]
```diff
@@ -107,23 +107,23 @@ func (bucket CouchbaseBucket) View(ddoc, name string, params map[string]interfac
 	return vres, bucket.Bucket.ViewCustom(ddoc, name, params, &vres)
 }
 
-func (bucket CouchbaseBucket) StartTapFeed(args sgbucket.TapArguments) (sgbucket.TapFeed, error) {
+func (bucket CouchbaseBucket) StartTapFeed(args sgbucket.TapArguments, notify sgbucket.BucketNotifyFn) (sgbucket.TapFeed, error) {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket.go]: Is there a reason to not use sgbucket.TapArguments to store/share the notify function?

File: src/github.com/couchbase/sync_gateway/base/bucket.go [hunk 2]
```diff
@@ -325,7 +330,7 @@ func (bucket CouchbaseBucket) CBSVersion() (major uint64, minor uint64, micro st
 }
 
 // Creates a Bucket that talks to a real live Couchbase server.
-func GetCouchbaseBucket(spec BucketSpec) (bucket Bucket, err error) {
+func GetCouchbaseBucket(spec BucketSpec, callback sgbucket.BucketNotifyFn) (bucket Bucket, err error) {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket.go:333]: Similar question here - is there a reason to not store the callback as part of the BucketSpec?

File: src/github.com/couchbase/sync_gateway/rest/changes_api.go [hunk 3]
```diff
@@ -218,6 +218,9 @@ func (h *handler) sendSimpleChanges(channels base.Set, options db.ChangesOptions
 			case <-closeNotify:
 				base.LogTo("Changes","Connection lost from client: %v", h.currentEffectiveUserName())
 				break loop
+			case <-h.db.ExitChanges:
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/rest/changes_api.go:221]: Does the close the underlying goroutine/feed that this loop is working?  It looks like timeout isn't sending a terminator to that feed either, but it feels like maybe it should?

File: src/github.com/couchbase/sync_gateway/rest/handler.go [hunk 4]
```diff
@@ -156,8 +170,26 @@ func (h *handler) invoke(method handlerMethod) error {
 		if err != nil {
 			return err
 		}
+
+		//If the Database is currently online process as normal
+		dbState := atomic.LoadUint32(&dbContext.State)
+		if (dbState == db.DBOnline) {
+			//Take a readlock for any method that does not run offline, run offline methods will take a write lock
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/rest/handler.go]: "run offline methods will take a write lock" - is this accurate?  Or do you just mean the online->offline transition takes the write lock?

File: src/github.com/couchbase/sync_gateway/rest/admin_api.go [hunk 5]
```diff
@@ -39,12 +43,94 @@ func (h *handler) handleCreateDB() error {
...
+	body, err := h.readBody()
+	if err != nil {
+		return err
+	}
+
+	var input struct {
+		Delay int           `json:"delay"`
+	}
+
+	input.Delay = kDefaultDBOnlineDelay;
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/rest/admin_api.go:69]: Do we have an option to set this delay in the config, as well as passing it to the online API?

File: src/github.com/couchbase/sync_gateway/rest/changes_api.go (hunk 3 above)
Comment (ajres) [inline src/github.com/couchbase/sync_gateway/rest/changes_api.go:221]: Feed is closed from  handleChanges with a defer close on the Terminator channel

```
options.Terminator = make(chan bool)
defer close(options.Terminator)
```

File: src/github.com/couchbase/sync_gateway/base/bucket.go (hunk 1 above)
Comment (ajres) [inline src/github.com/couchbase/sync_gateway/base/bucket.go]: notify function is now passed in the sgbucket.TapArguments

File: src/github.com/couchbase/sync_gateway/base/bucket.go (hunk 2 above)
Comment (ajres) [inline src/github.com/couchbase/sync_gateway/base/bucket.go:333]: notify function is now passed in the sgbucket.TapArguments

File: src/github.com/couchbase/sync_gateway/rest/handler.go (hunk 4 above)
Comment (ajres) [inline src/github.com/couchbase/sync_gateway/rest/handler.go]: I have removed this generic statement, it is the _offline code that takes the write lock to block on existing open calls

File: src/github.com/couchbase/sync_gateway/rest/admin_api.go (hunk 5 above)
Comment (ajres) [inline src/github.com/couchbase/sync_gateway/rest/admin_api.go:69]: delay values in server config is not currently supported. The DB can be started in offline mode, then brought online with a delay via the REST API.

File: src/github.com/couchbase/sync_gateway/rest/handler.go [hunk 6]
```diff
@@ -156,8 +170,29 @@ func (h *handler) invoke(method handlerMethod) error {
...
+
+				//defer releasing the dbContext until after the handler method returns
+				defer dbContext.AccessLock.RUnlock()
+			}
+		} else if (dbState == db.DBOffline) { //DB is offline
+			if (!h.runOffline) { //only handlers with runOffline true can run if DB is in Offline state
+				return base.HTTPErrorf(http.StatusServiceUnavailable, "DB is currently under maintenance")
+			}
+		} else { //DB is in transition state, no calls will be accepted until it is Online or Offline state
+			return base.HTTPErrorf(http.StatusServiceUnavailable, "DB is currently in neither online or offline state, try again later")
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/rest/handler.go]: This error message sounds a bit too Alice in Wonderland - maybe just say "Database is not online - try again later"?

Comment (ajres) [inline src/github.com/couchbase/sync_gateway/rest/handler.go]: This is also designed to stop e.g. _resync if the DB is not Offline.

Maybe "Database's state is transitioning - try again later"?

File: src/github.com/couchbase/sync_gateway/rest/handler.go [hunk 7]
```diff
@@ -156,8 +170,29 @@ func (h *handler) invoke(method handlerMethod) error {
...
 		}
+
+		//If the Database is currently online process as normal
+		dbState := atomic.LoadUint32(&dbContext.State)
+		if (dbState == db.DBOnline) {
+			//Take a readlock for any method that does not run offline, taking a DB offline will
+			//block until all normal calls have returned and released their read locks.
+			if (!h.runOffline) {
+				//get a read lock on the dbContext
+				dbContext.AccessLock.RLock()
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/rest/handler.go]: It feels like there's a potential race condition here, in a scenario like:
1. This code completes the evaluation on line 176.
2. TakeDbOffline is called, updates state to stopping and blocks waiting for the write lock
3. This code gets to line 181, and blocks here trying to get the RLock 
4. After the write lock is released, this request proceeds, even though the DB is offline.

File: src/github.com/couchbase/sync_gateway/rest/handler.go (hunk 6 above)
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/rest/handler.go]: Maybe we should define strings and report the actual state?  (e.g. stopping, starting, etc)

**Comment by ajres** [comment]
With reference to performance impact of taking a read lock on all REST API calls, these are some golang benchmark tests run with 1,2,4 cpus. Test was calling GET on a single document, the middle column is the number of iterations run in the benchmark timeof one minute.

online_offline

Benchmark_RestApiGetDocPerformance         50000       1495821 ns/op
Benchmark_RestApiGetDocPerformance-2      100000        993498 ns/op
Benchmark_RestApiGetDocPerformance-4      100000        751594 ns/op

Benchmark_RestApiGetDocPerformance                 50000       1558559 ns/op
Benchmark_RestApiGetDocPerformance-2      100000        957619 ns/op
Benchmark_RestApiGetDocPerformance-4      200000        706654 ns/op

master

Benchmark_RestApiGetDocPerformance            100000       1462693 ns/op
Benchmark_RestApiGetDocPerformance-2      100000        976649 ns/op
Benchmark_RestApiGetDocPerformance-4      200000        712329 ns/op

Benchmark_RestApiGetDocPerformance          50000      1407834 ns/op
Benchmark_RestApiGetDocPerformance-2      100000        945054 ns/op
Benchmark_RestApiGetDocPerformance-4      200000        700864 ns/op

**Comment by ajres** [comment]
latest jenkins [dev build](http://mobile.jenkins.couchbase.com/view/Sync_Gateway_Dev/job/sync-gateway-dev-build/381/)

**Pull request merged**